### PR TITLE
driver/sensor: add LIS2DW12 sensor support

### PR DESCRIPTION
**Pull request by avisconti**
Add support to STM LIS2DW12 3-axis accelerometer driver.
The driver support I2C and SPI bus communication and both
polling and drdy trigger mode.

Co-authored-by: Mario Tesi <mario.tesi@st.com>
Signed-off-by: Armando Visconti <armando.visconti@st.com>

This PR replaces #7627

EDIT::
tested in I2C with component on IKS01A3 shield on top of Nucleo F401RE board
tested in SPI with external component mounted on DIL24 socket of same shield
In both cases the interrupt triggered mode was used.
In case of SPI the CS was generated using a GPIO.
(Please note that IKS01A3 shield is not officially supported yet. I'm still working on it)

### CODE REVIEW

**Comment by avisconti** [comment]
rebased and repushed

**Comment by avisconti** [comment]
Replaced definitions for CS_GPIO with the ones generated by DTS.

**Comment by avisconti** [comment]
recheck

**Comment by avisconti** [comment]
fixed couple of typos

File: drivers/sensor/CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +23,7 @@ add_subdirectory_ifdef(CONFIG_ISL29035		isl29035)
 add_subdirectory_ifdef(CONFIG_LIS2DH		lis2dh)
 add_subdirectory_ifdef(CONFIG_LIS2DS12		lis2ds12)
 add_subdirectory_ifdef(CONFIG_LIS2MDL		lis2mdl)
+add_subdirectory_ifdef(CONFIG_LIS2DW12          lis2dw12)
```
Comment (MaureenHelm) [inline drivers/sensor/CMakeLists.txt]: Move this up one line to preserve alpha order

File: drivers/sensor/Kconfig [hunk 2]
```diff
@@ -75,6 +75,8 @@ source "drivers/sensor/lis2ds12/Kconfig"
 
 source "drivers/sensor/lis2mdl/Kconfig"
 
+source "drivers/sensor/lis2dw12/Kconfig"
```
Comment (MaureenHelm) [inline drivers/sensor/Kconfig]: Move this up one line to preserve alpha order

File: drivers/sensor/lis2dw12/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+zephyr_library()
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/CMakeLists.txt]: Add license/copyright header

File: drivers/sensor/lis2dw12/lis2dw12.c [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		lis2dw12_channel_get_acc(dev, chan, val);
+		return 0;
+	default:
+		SYS_LOG_DBG("Channel not supported");
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/lis2dw12.c]: This should be LOG_DBG

File: drivers/sensor/lis2dw12/lis2dw12.c [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+			    enum sensor_attribute attr,
+			    const struct sensor_value *val)
+{
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		return lis2dw12_set_range(dev, sensor_ms2_to_g(val));
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return lis2dw12_set_odr(dev, val->val1);
+	default:
+		SYS_LOG_DBG("Acc attribute not supported");
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/lis2dw12.c]: This should be LOG_DBG

File: drivers/sensor/lis2dw12/lis2dw12.c [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+	/* adjust to resolution */
+	if (cfg->pm == LIS2DW12_LOW_POWER_M1) {
+		shift = LIS2DW12_SHIFT_PM1;
+	} else {
+		shift = LIS2DW12_SHIFT_PMOTHER;
+	}
+
+	lis2dw12->acc[0] = sys_le16_to_cpu(buf.a_axis[0] >> shift);
+	lis2dw12->acc[1] = sys_le16_to_cpu(buf.a_axis[1] >> shift);
+	lis2dw12->acc[2] = sys_le16_to_cpu(buf.a_axis[2] >> shift);
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/lis2dw12.c]: I think you need to do the shift after the byte swap?

File: drivers/sensor/lis2dw12/lis2dw12.c [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+				    LIS2DW12_SHFT_GAIN_NOLP1 : 0);
+
+#ifdef CONFIG_LIS2DW12_TRIGGER
+	if (lis2dw12_init_interrupt(dev) < 0) {
+		SYS_LOG_ERR("Failed to initialize interrupts");
+		return -EIO;
+	}
+#endif /* CONFIG_LIS2DW12_TRIGGER */
+
+	dev->driver_api = &lis2dw12_driver_api;
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/lis2dw12.c]: Remove this line, and replace `DEVICE_INIT` with `DEVICE_AND_API_INIT` below

File: drivers/sensor/lis2dw12/lis2dw12.h [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+#ifndef ZEPHYR_DRIVERS_SENSOR_LIS2DW12_LIS2DW12_H_
+#define ZEPHYR_DRIVERS_SENSOR_LIS2DW12_LIS2DW12_H_
+
+#include <gpio.h>
+#include <misc/util.h>
+#include <sensor.h>
+
+#define SYS_LOG_DOMAIN "LIS2DW12"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_SENSOR_LEVEL
+#include <logging/sys_log.h>
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/lis2dw12.h]: Need to switch to `logging/log.h`

File: drivers/sensor/lis2dw12/lis2dw12.h [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+	const char *bus_name;
+	enum lis2dh_powermode pm;
+#ifdef CONFIG_LIS2DW12_TRIGGER
+	const char *int_gpio_port;
+	u8_t int_gpio_pin;
+	u8_t int_pin;
+#endif /* CONFIG_LIS2DW12_TRIGGER */
+};
+
+struct lis2dw12_data;
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/lis2dw12.h]: Why is this here?

File: drivers/sensor/lis2dw12/lis2dw12_spi.c [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+#ifdef DT_ST_LIS2DW12_BUS_SPI
+
+#define LIS2DW12_SPI_READ		(1 << 7)
+
+#define LOG_LEVEL CONFIG_SENSOR_LOG_LEVEL
+LOG_MODULE_DECLARE(LIS2DW12);
+
+#if defined(DT_ST_LIS2DW12_0_CS_GPIO_CONTROLLER)
+static struct spi_cs_control lis2dw12_cs_ctrl;
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/lis2dw12_spi.c]: Why aren't you using the .cs field in struct spi_config?

File: drivers/sensor/lis2dw12/lis2dw12_trigger.c [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+int lis2dw12_init_interrupt(struct device *dev)
+{
+	struct lis2dw12_data *lis2dw12 = dev->driver_data;
+	const struct lis2dw12_device_config *cfg = dev->config->config_info;
+	int ret;
+
+	/* setup data ready gpio interrupt (INT1 or INT2) */
+	lis2dw12->gpio = device_get_binding(cfg->int_gpio_port);
+	if (lis2dw12->gpio == NULL) {
+		SYS_LOG_DBG("Cannot get pointer to %s device",
```
Comment (MaureenHelm) [inline drivers/sensor/lis2dw12/lis2dw12_trigger.c]: This should be LOG_DBG

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/Kconfig (hunk 2 above)
Comment (avisconti) [reply drivers/sensor/Kconfig]: ack

File: drivers/sensor/lis2dw12/CMakeLists.txt (hunk 3 above)
Comment (avisconti) [reply drivers/sensor/lis2dw12/CMakeLists.txt]: Very few CMakelist.txt have the header, but I agree that they should.
I'll add it also to LPS22HH, which is the other sensor driver I PR-ed recently

File: drivers/sensor/lis2dw12/lis2dw12.c (hunk 4 above)
Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12.c]: OK, I'll correct all LOG part globally.

File: drivers/sensor/lis2dw12/lis2dw12.h (hunk 9 above)
Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12.h]: Because lis2dw12_data and lis2dw12_tf structures are referrring one to each other. So, in order to compile properly, I need to declare lis2dw12_data before lis2dw12_tf, and then define all the
fields properly after. It is a common practice done also in other sensor drivers too
(LSM6DSL and LIS2DS12)

File: drivers/sensor/lis2dw12/lis2dw12_spi.c (hunk 10 above)
Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12_spi.c]: Actually this is the driver istance of the spi_cs_control. Its address goes in .cs and is used by 
SPI driver.
If DT_ST_LIS2DW12_0_CS_GPIO_CONTROLLER is not defined then .cs remains NULL
and the SPI chip select will be generated by the controller itself.

File: drivers/sensor/lis2dw12/lis2dw12_trigger.c (hunk 11 above)
Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12_trigger.c]: sure. All the LOG related changes will be applied in the PR update.

**Comment by avisconti** [comment]
@MaureenHelm 
re-tested and re-pushed.
Thanks for the review!

File: drivers/sensor/lis2dw12/CMakeLists.txt (hunk 3 above)
Comment (MaureenHelm) [reply drivers/sensor/lis2dw12/CMakeLists.txt]: Thanks!

File: drivers/sensor/lis2dw12/lis2dw12.c (hunk 6 above)
Comment (MaureenHelm) [reply drivers/sensor/lis2dw12/lis2dw12.c]: Please check this

File: drivers/sensor/lis2dw12/lis2dw12.h (hunk 9 above)
Comment (MaureenHelm) [reply drivers/sensor/lis2dw12/lis2dw12.h]: It wasn't initially obvious that this was a forward declaration. It's helpful to comment as such.

File: drivers/sensor/lis2dw12/lis2dw12_spi.c (hunk 10 above)
Comment (MaureenHelm) [reply drivers/sensor/lis2dw12/lis2dw12_spi.c]: My mistake, I missed that this was .cs was a pointer. That said, why isn't the instance part of `struct lis2dw12_data`?

File: drivers/sensor/lis2dw12/lis2dw12.h (hunk 9 above)
Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12.h]: OK, I'm adding a comment.

File: drivers/sensor/lis2dw12/lis2dw12_spi.c (hunk 10 above)
Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12_spi.c]: OK, I can put it in data structure. It seems cleaner to me.

File: drivers/sensor/lis2dw12/lis2dw12.c (hunk 6 above)
Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12.c]: This is a piece of code which I didn't pay attention to much. It was written by Mario and I
re-used it as it is (and it worked fine). 
But let me give a closer look to it and your observations.

Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12.c]: OK, let me add some comments (even for myself):
1. The reason why is working on my h/w setup is because the cpu is Little Endian, so the
    swap macro does not actually swap.
2. I think you are right. If the driver is running on a Big Endian platform the driver has to transform
    data into cpu order, and only then perform the shift. 
    So I'll correct this, thx!!!
3. I just realize that some of ST driver does not have this swap inside, so the won't work correctly
    on Big Endian platforms. I will submit a proper fix. Thx!

**Review by MaureenHelm — APPROVED** [review]
